gateway/session_manager.go

```diff
@@ -333,97 +333,128 @@ func (l *SessionLimiter) ForwardMessage(r *http.Request, session *user.SessionSt
 
 // RedisQuotaExceeded returns true if the request should be blocked as over quota.
 func (l *SessionLimiter) RedisQuotaExceeded(r *http.Request, session *user.SessionState, quotaKey, scope string, limit *user.APILimit, store storage.Handler, hashKeys bool) bool {
+	logger := log.WithFields(logrus.Fields{
+		"quotaMax":         limit.QuotaMax,
+		"quotaRenewalRate": limit.QuotaRenewalRate,
+	})
+
 	if limit.QuotaMax <= 0 {
+		logger.Error("Quota disabled: quota max <= 0")
 		return false
 	}
 
 	// don't use the requests cancellation context
 	ctx := context.Background()
 
-	session.Touch()
-
 	quotaScope := ""
 	if scope != "" {
 		quotaScope = scope + "-"
 	}
 
 	key := session.KeyID
-
 	if hashKeys {
 		key = storage.HashStr(session.KeyID)
 	}
-
 	if quotaKey != "" {
 		key = quotaKey
 	}
 
 	now := time.Now().Truncate(0)
+
+	// rawKey is the redis key for quota
 	rawKey := QuotaKeyPrefix + quotaScope + key
-	quotaRenewalRate := time.Second * time.Duration(limit.QuotaRenewalRate)
-	quotaMax := limit.QuotaMax
 
-	// First, ensure a distributed lock
+	quotaRenewalRate := time.Second * time.Duration(limit.QuotaRenewalRate)
 	conn := l.limiterStorage
-	locker := limiter.NewLimiter(conn).Locker(rawKey)
 
-	if err := locker.Lock(ctx); err != nil {
-		log.WithError(err).Error("error locking quota key, blocking")
+	var expired, exists bool
+	var expiredAt time.Time
+
+	//logger = logger.WithField("key", rawKey)
+
+	dur, err := conn.PTTL(ctx, rawKey).Result()
+	if err != nil && !errors.Is(err, redis.Nil) {
+		logger.WithError(err).Error("error getting key TTL, blocking")
 		return true
 	}
-	defer func() {
-		if err := locker.Unlock(ctx); err != nil {
-			log.WithError(err).Error("error unlocking quota key")
+
+	// The command returns -2 if the key does not exist.
+	// The command returns -1 if the key exists but has no associated expire.
+	expired = dur < 0
+	exists = dur != -2
+
+	expiredAt = now.Add(dur)
+
+	logger = logger.WithFields(logrus.Fields{
+		"exists":  exists,
+		"expired": expired,
+		"rawKey":  rawKey,
+		//		"expiredAt": expiredAt,
+	})
+
+	increment := func() bool {
+		var res *redis.IntCmd
+		conn.Pipelined(ctx, func(pipe redis.Pipeliner) error {
```

Comment:
does this need to be a transaction?

Reply:
It does not. ExpireNX is conditional (set ttl only if ttl not set), and Incr is atomic on redis side;